cloud-init: 0.7.9 -> 20.2 (python3!)

Mic92 opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

flokli committed:
nixos/cloud-init: actually test ssh'ing, also, with cloud-init-created user

Mic92 committed:
cloud-utils: split of smaller .guest output

Mic92 committed:
cloud-init: 0.7.9 -> 20.2

Mic92 commented on nixos/tests/cloud-init.nix:
```diff
@@ -18,13 +18,15 @@ let
       cat << EOF > $out/iso/meta-data
       instance-id: iid-local01
       local-hostname: "test"
-      public-keys:
```
No idea why this is no longer picked up. The schema of this file is unclear.

Mic92 replied:
works!

Mic92 commented on pkgs/tools/virtualization/cloud-init/default.nix:
```diff
@@ -1,40 +1,83 @@
+  ];
 
-  checkInputs = with pythonPackages; [ contextlib2 httpretty mock unittest2 ];
+  disabledTests = [
```
Almost all tests are running!

Mic92 commented on pkgs/tools/virtualization/cloud-init/default.nix:
```diff
@@ -1,40 +1,83 @@
   };
 
-  patches = [ ./add-nixos-support.patch ];
+  patches = [ ./0001-add-nixos-support.patch ];
```
I think about upstreaming this.

flokli replied:
I think we should - the patch is generic enough for it.

zimbatm commented on pkgs/tools/virtualization/cloud-init/0001-add-nixos-support.patch:
```diff
@@ -1,18 +1,36 @@
+@@ -46,6 +46,7 @@ OSFAMILIES = {
      'freebsd': ['freebsd'],
-     'suse': ['sles'],
+     'suse': ['opensuse', 'sles'],
```
is there a reason to add opensuse?

Mic92 replied:
I copied over the old patch. Must be a left-over. Will remove.

Mic92 replied:
Actually github is confusing here. The new version here has `opensuse` here. This is not a change that I made.

zimbatm commented on pkgs/tools/virtualization/cloud-init/0001-add-nixos-support.patch:
```diff
@@ -72,7 +95,7 @@ diff -ruN cloud-init-0.7.6.orig/cloudinit/distros/nixos.py cloud-init-0.7.6/clou
 +        if not conf:
 +            conf = HostnameConf('')
 +        conf.set_hostname(your_hostname)
-+        util.write_file(out_fn, str(conf), 0644)
```
is 0644 the default mode if left unspecified?

Mic92 replied:
It will use the permission of the old file, which is `.r--r--r-- 14 root  1 Jan  1970  /nix/store/g1mkk8lygngbzvm632wbmnm8xxf9dymz-etc-hostname
`

Mic92 replied:
Same here. The actual line you are interested in is below.

zimbatm commented on pkgs/tools/virtualization/cloud-init/0001-add-nixos-support.patch:
```diff
@@ -52,6 +73,8 @@ diff -ruN cloud-init-0.7.6.orig/cloudinit/distros/nixos.py cloud-init-0.7.6/clou
 +        # should only happen say once per instance...)
 +        self._runner = helpers.Runners(paths)
++        self.usr_lib_exec = os.path.join(os.path.dirname(__file__),
++                                         "../../../../../libexec")
```
where does that point to, the current-system libexec?

Mic92 replied:
`/usr/lib` if not overwritten. Not it points to the `$out/libexec`.

zimbatm commented on pkgs/tools/virtualization/cloud-init/default.nix:
```diff
@@ -1,40 +1,83 @@
+    owner = "canonical";
+    repo = "cloud-init";
+    rev = version;
+    sha256 = "sha256-QeY/fdIIUSsp5oNxyRtZwpTB747Jf5KAJuYY9yiKUvc=";
```
are SRI hashes allowed now?

zimbatm reviewed (commented):
nice! This should be included in 20.09 if possible.

Mic92 commented:
@zimbatm could you give this a shot on ec2 or gcp. Or can you refer to someone else who can test it?

zimbatm commented:
maybe @rbvermaa ?

flokli commented:
@Mic92 I just stubled over some old work from me in January on this (which I never pushed, boo).

I just rebased and pushed it to https://github.com/flokli/nixpkgs/commits/cloud-init.

At least https://github.com/flokli/nixpkgs/commit/b19bcea4089f6292be44c6266eca18d95fb6d533 seems to be a good addition for more test coverage, and maybe the `hostnamectl` patch as well, if applicable to the more recent `20.2` version (given NixOS now support setting transient ones).

Mic92 commented:
@flokli feel free to push it on top. I won't get back to until 9th october.

flokli commented:
I was also quite busy spending time away from computers :laughing:

Hope you're fine - let me know if you want to pick this up. I currently don't really have any strong ties with cloud-init, but felt like it'd be a waste to just ignore these commits.

flokli committed:
nixosTests.cloud-init: test changing hostname actually works

flokli committed:
cloud-init: 20.2 -> 20.3

required rebasing the patch, disabling some tests.

I also changed the hash to be in conventional format - the

hash mismatch in fixed-output derivation '/nix/store/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx-source':
  wanted: sha256:yyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyy
  got:    sha256:zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz

doesn't propose SRI syntax.

flokli commented:
Pushed my commits on top.

I opted against changing the cloud-init patch to use `hostnamectl`, because:
 - This would require dropping  https://github.com/NixOS/nixpkgs/blob/master/pkgs/os-specific/linux/systemd/default.nix#L45. Currently `hostnamectl set-hostname`… throws a NixOS-specific error message.
 - Even when I drop this patch, it only seemed to affect the kernel hostname retrieved via `uname -n`, not the output of the `hostname` command. I'm not sure if this is intended behaviour, but just changing the hostname like it was in the patch did work, so fine for me.

I also added a test checking the hostname change being successful, PTAL.

zimbatm reviewed (approved):
LGTM
